### ISSUE TITLE
Numpy is ignoring the fact "isfinite" is undefined (migrated from Trac #1625)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3176]
Original ticket http://projects.scipy.org/numpy/ticket/1625
Reported 2010-10-05 by trac user drkirkby, assigned to unknown.

On Solaris or !OpenSolaris, when Python is built, it is not compiled as a C99 application. (That's a choice the Python developers took). 

Therefore, {{{isfinite}}}, which was not defined in the C standard until C99, does not exist. Hence when Python builds, the config.log shows:

```
ac_cv_have_decl_isfinite=no
```

The latest Numpy, (and this includes the latest snapshot taken only an hour or two) ago builds OK on Solaris. However, when one tries to import this into Python with:

```
import numpy
```

it generates an error:

```
ImportError: ld.so.1: python: fatal: relocation error: file /home/palmieri/fulvia/numpy/sage-4.6.alpha2/local/lib/python2.6/site-packages/numpy/core/multiarray.so: symbol isfinite: referenced symbol not found
```

It would appear that Numpy is trying to use {{{isfinite}}}, despite the fact that Python is built without adding specific C99 support. 

Linux systems, which I assume is what most Numpy developers use, defines {{{isfinite}}} when one includes {{{math.h}}} in a file. But this should not happen unless one builds C99 code, as {{{isfinite}}} was not defined until the C99 standard. That's a  problem with Linux - it tends not to conform to standards too well. 

Solaris in comparison will not define {{{isfinite}}} unless the code is C99. {{{isfinite}}} is defined in a header file {{{/usr/include/iso/math_c99.h}}} but '''not''' in {{{/usr/include/math.h}}}. However, if the compiler is building C99 code, then the contents of {{{/usr/include/iso/math_c99.h}}} will be included, so {{{isfinite}}} will be defined. 

I believe this is a bug in Numpy, as it should respect the fact the Python build has determined {{{isfinite}}} is undefined (so sets {{{ac_cv_have_decl_isfinite=no}}}) and so therefore Numpy should not try to use {{{isfinite}}}. 

This came up as a bug when we tried to update Numpy in Sage. It is documented on 

http://trac.sagemath.org/sage_trac/ticket/9808

but I believe you will find the information there confusing, as it covers many issues associated with upgrading Numpy and Scipy in Sage. 

The basic problem is Numpy is using {{{isfinite}}} even though Python has worked out that it is undefined. 

Dave

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2010-10-05

The STSI folks haven't reported this on SPARC, so this may be a hardware/version/compiler specific problem. Can you provide some details about those? And what python version are you using?

On my system that the relevant lines are in pyconfig.h

```
/* Define to 1 if you have the declaration of `isfinite', and to 0 if you
   don't. */
#define HAVE_DECL_ISFINITE 1
```

And HAVE_DECL_ISFINITE is what Numpy looks for. Where does ac_cv_have_decl_isfinite come from?

Note that it is not a bug to provide more than a standard requires.

**Comment by thouis** [comment]
Comment in Trac by trac user drkirkby, 2010-10-05

Replying to [comment:1 charris]:

> The STSI folks haven't reported this on SPARC, so this may be a hardware/version/compiler specific problem. Can you provide some details about those? And what python version are you using?

Sorry, I should have been more specific. This problem is seen with gcc 4.5.1, using Python 2.6.4. It's seen on three systems. These include both SPARC and x86. 
- Sun Blade 2500 SPARC running Solaris 10 update 1 (Jan 2006 edition)
- Dell server, Xeon CPUs, Solaris 10 update 5 (May 2005 release)
- Sun Ultra 27, quad core Xeon running !OpenSolaris 06/09 (June 2009 release)

> On my system that the relevant lines are in pyconfig.h
> 
> {{{
> /\* Define to 1 if you have the declaration of `isfinite', and to 0 if you
>    don't. */
> #define HAVE_DECL_ISFINITE 1
> }}}
> 
> And HAVE_DECL_ISFINITE is what Numpy looks for. Where does ac_cv_have_decl_isfinite come from?

{{{ac_cv_have_decl_isfinite}}} is in Python's config.log. Python's header file {{{pyconfig.h}}} has in it:

```
/* Define to 1 if you have the declaration of `isfinite', and to 0 if you
   don't. */
#define HAVE_DECL_ISFINITE 0
```

when Numpy 1.5.0 is built, there are some compiler warnings, one of which is:

```
/src.solaris-2.11-i86pc-2.6/numpy/core/include/numpy/
buildconfig.h:101:0: warning: "HAVE_DECL_ISFINITE" redefined
/export/home/drkirkby/sage-4.6.alpha2/local/include/python2.6/pyconfig.h:121:0: note: this is the location of the previous definition
```

Numpy's file {{{config.h}}} has in it:

```
#define HAVE_DECL_ISINF
#define HAVE_DECL_ISFINITE
#define HAVE_COMPLEX_H
```

Of course, it is true that {{{HAVE_DECL_ISFINITE}}} was defined in Python's {{{pyconfig.h}}}, but it was defined to a value of {{{0}}} indicating the system did '''not''' have {{{isfinite}}}. 

> Note that it is not a bug to provide more than a standard requires.

I take your point, but it does cause problems when people write code and assume that they can use things like {{{isfinite}}} as they are declared in {{{math.h}}}, when in fact they are not standard. This seems to be the case in point here. 

Dave

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2010-10-05

> I take your point, but it does cause problems when people write code and 
> assume that they can use things like isfinite as they are declared in math.h,
> when in fact they are not standard. This seems to be the case in point here.

That is not the issue here. The build does check whether `isfinite`
is available.

However, the code at `numpy/core/setup.py:218` is wrong, since it checks
for `#ifdef` and not `#if` for Python's `HAVE_DECL_ISFINITE`.

**Comment by thouis** [comment]
Comment in Trac by trac user drkirkby, 2010-10-05

Thank you. 

Can you patch this for 1.5.1? When is that likely to be released? (Very approximately). 

Dave

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2010-10-05

''This seems to be the case in point here.'' 

I don't think so. I suspect it is this bit of code in numpy/distutils/command/config.py

```
int main()
{
#ifndef %s
    (void) %s;
#endif
    ;
    return 0;
}""" % (symbol, symbol)
```

which only looks to see if the macro is defined, not whether it is true of false. I suspect we need to make the compilation fail if the symbol is defined to 0. Maybe changing it to

```
int main()
{
#ifndef %s
    (void) %s;
#elif %s == 0
#error
#endif
    ;
    return 0;
}""" % (symbol, symbol, symbol)
```

will do the trick. The numpy build process is pretty complicated so it would be nice if David Cournapeau could take a look at this.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2010-10-05

Suggested fix:

http://github.com/pv/numpy/compare/master...bug/1625-isfinite-decl.diff

Testing would be useful, I don't have a machine at hand without C99.

We can fix it for 1.5.1, but at the moment there's no schedule for that release. It'll probably appear, since there were some Python 3 bugs still left in 1.5.0, and since 2.0.0 will break binary compatibility.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2010-10-05

I think we do need a 1.5.1, especially for projects like SAGE that need some of the new features but are unlikely to move to 2.0.0 anytime soon. Same for ditros like Fedora, which doesn't seem to track Numpy very closely.

**Comment by thouis** [comment]
Comment in Trac by trac user drkirkby, 2010-10-06

Replying to [comment:6 pv]:

> Suggested fix:
> 
> http://github.com/pv/numpy/compare/master...bug/1625-isfinite-decl.diff
> 
> Testing would be useful, I don't have a machine at hand without C99.

It does appear to work on Solaris x86 and !OpenSolaris. We have not checked on SPARC, but I'm pretty confident this patch does work. Someone said ''it would be nice if David Cournapeau could take a look at this. '' Could he be cc'ed on the ticket, so he is aware of it, and hopefully comment on whether the patch is the right way to do it? I don't know his details, but I assume someone here does. 

> We can fix it for 1.5.1, but at the moment there's no schedule for that release. It'll probably appear, since there were some Python 3 bugs still left in 1.5.0, and since 2.0.0 will break binary compatibility.

OK. Thank you. It would be good if there was one, as there are a few patches now in Sage to get around bugs which have been fixed in Numpy. 

Dave

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2010-10-06

Applied in commit:5a0ab02 and commit:8e0d299
